mqtt/mqttd/src/app/edgehub.rs

```diff
@@ -0,0 +1,285 @@
+
+                // awaits for at least one to finish
+                let (_res, _stopped, sidecars) = sidecars.await;
+
+                // wait for the rest to exit
+                future::join_all(sidecars).await;
+
+                snapshot?
+            }
+            // one of sidecars exited first
+            Either::Right(((res, stopped, sidecars), server)) => {
+                // signal server
+                broker_handle.send(Message::System(SystemEvent::Shutdown))?;
+                let snapshot = server.await;
+
+                debug!("a sidecar has stopped. shutting down all sidecars...");
+                if let Err(e) = res {
+                    error!(message = "failed waiting for sidecar shutdown", error = %e);
+                }
+
+                // send shutdown event to each of the rest sidecars
+                shutdowns.remove(stopped);
+                let shutdowns = shutdowns.into_iter().map(SidecarShutdownHandle::shutdown);
+                future::join_all(shutdowns).await;
+
+                // wait for the rest to exit
+                future::join_all(sidecars).await;
+
+                snapshot?
+            }
+        };
+
+        Ok(state)
+    }
+}
+
+async fn make_server<Z>(
+    config: Settings,
+    broker: Broker<Z>,
+    broker_ready: BrokerReady,
+) -> Result<Server<Z, MakeEdgeHubPacketProcessor<MakeMqttPacketProcessor>>>
+where
+    Z: Authorizer + Send + 'static,
+{
+    let broker_handle = broker.handle();
+
+    let make_processor = MakeEdgeHubPacketProcessor::new_default(broker_handle.clone());
+    let mut server = Server::from_broker(broker).with_packet_processor(make_processor);
+
+    // Add system transport to allow communication between edgehub components
+    let authenticator = LocalAuthenticator::new();
+    server.with_tcp(config.listener().system().addr(), authenticator, None)?;
+
+    // Add regular MQTT over TCP transport
+    let authenticator = EdgeHubAuthenticator::new(config.auth().url());
+
+    if let Some(tcp) = config.listener().tcp() {
+        let broker_ready = Some(broker_ready.signal());
+        server.with_tcp(tcp.addr(), authenticator.clone(), broker_ready)?;
+    }
+
+    // Add regular MQTT over TLS transport
+    if let Some(tls) = config.listener().tls() {
+        let identity = if let Some(config) = tls.certificate() {
+            info!("loading identity from {}", config.cert_path().display());
+            ServerCertificate::from_pem(config.cert_path(), config.private_key_path())
+                .with_context(|| {
+                    ServerCertificateLoadError::File(
+                        config.cert_path().to_path_buf(),
+                        config.private_key_path().to_path_buf(),
+                    )
+                })?
+        } else {
+            info!("downloading identity from edgelet");
+            download_server_certificate()
+                .await
+                .with_context(|| ServerCertificateLoadError::Edgelet)?
```

Comment:
as briefly discussed offline, we need to make sure we capture and display source of errors as well.